docs/source/features/memory_optimizations.rst

```diff
@@ -5,6 +5,80 @@ Parallelism
 -----------
 Refer to :doc:`Parallelism <./parallelisms>`.
 
+
+Mixture of Experts
+------------------
+
+Overview
+^^^^^^^^
+
+NeMo supports Mixture of Experts (MoE)in the transformer layer for NLP models.
+
+MoE is a machine learning technique where multiple specialized models (experts,
+usually multi-layer perceptrons) are combined to solve a complex task. Each expert
+focuses on a specific subtask or domain, while a gating network dynamically activates
+the most appropriate expert based on the current input.
+
+
+To use MoE  in the NeMo Framework, adjust the ``num_moe_experts`` parameter in the model configuration:
+
+1. Set ``num_moe_experts`` to `8` to leverage 8 experts in the MoE module.
+
+   .. code-block:: yaml
+
+       num_moe_experts: 8  # Set MoE to use 8 experts
+
+2. Set ``moe_router_topk`` to the number of experts you want activated, for eaxmple if you want to process each input with two experts:
+
+   .. code-block:: yaml
+
+       moe_router_topk: 2  # Processes each token using 2 experts.
+
+In addition NeMo provides options to configure MoE-specific loss function, for example to balance token
+distribution across experts:
+
+1. Set ``moe_router_load_balancing_type`` to specify the load balancing method:
+
+   .. code-block:: yaml
+
+      moe_router_load_balancing_type: aux_loss  # to use the auxilary loss, other options include "sinkhorn".
+
+2. Set ``moe_aux_loss_coeff`` to specify the weight of the auxilary loss, values in the 1e-2 range are a good start:
+
+   .. code-block:: yaml
+
+      moe_aux_loss_coeff: 1e-2  # set the aux-loss weight to 1e-2
+
+3. Set ``moe_z_loss_coeff`` to specify the weight of the z-loss, a starting value of 1e-3 is recommended:
+
+   .. code-block:: yaml
+
+      moe_z_loss_coeff: 1e-3
+
+Other options include:
+
+1. ``moe_input_jitter_eps`` adds noise to the input tensor by applying jitter with a specified epsilon value
+
+2. ``moe_token_dropping`` enables selectively dropping and padding tokens for each expert to achieve
+   a specified capacity.
+
+3. ``moe_token_dropping`` specifies the token dispatcher type, options include 'allgather' and 'alltoall'.
+
+4. ``moe_per_layer_logging`` enables per-layer logging for MoE, currently support aux-loss and z-loss.
+
+5. ``moe_expert_capacity_factor`` the capacity factor for each expert, None means no token will be dropped. The default is None.
+
+6. ``moe_pad_expert_input_to_capacity`` if True, pads the input for each expert to match the expert capacity length, effective only after the moe_expert_capacity_factor is set. The default setting is False.
+
+7. ``moe_token_drop_policy`` the policy to drop tokens. Can be either "probs" or "position". If "probs", the tokens with the lowest probabilities will be dropped. If "position", tokens at the end of each batch will be dropped. Default value is "probs".
+
+8. ``moe_layer_recompute`` if True, checkpointing moe_layer to save activation memory, default is False.
+
```

Comment:
Additional copyedits in the file.


**Line 181 Add period at end of sentence.**

Check implementation details from Attention Class in Megatron Core Repo: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/attention.py#L49.


**Lines 192 - 196. Bad formatting. The > character collapses the lines into a single paragraph. Remove > and replace with the rst bullet character *.**

* Supports training models with long sequence lengths by managing activation memory efficiently.
* Enables high batch sizes per GPU by offloading activation memory.
* Overlaps computation with data transfers (Host2Device and Device2Host) during offloading and reloading.

**Lines 200 - 202. Bad formatting. The > character collapses the lines into a single paragraph. Remove > and replace with the rst bullet character *..**

* Set cpu_offloading to True to enable CPU offloading.
* Set cpu_offloading_num_layers to a value between 0 and the total number of layers in the model minus one.
* Set cpu_offloading_activations and cpu_offloading_weights based on your needs to offload activations only, weights only, or both.